cmd/createorupdate/update.go

```diff
@@ -0,0 +1,168 @@
+package main
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/log"
+)
+
+func update(cs *api.OpenShiftManagedCluster, p api.Plugin) error {
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return err
+	}
+
+	ssc := compute.NewVirtualMachineScaleSetsClient(cs.Properties.AzProfile.SubscriptionID)
+	ssc.Authorizer = authorizer
+	vmc := compute.NewVirtualMachineScaleSetVMsClient(cs.Properties.AzProfile.SubscriptionID)
+	vmc.Authorizer = authorizer
+
+	ctx := context.Background()
+
+	err = updateInPlace(ctx, cs, p, ssc, vmc, api.AgentPoolProfileRoleMaster)
+	if err != nil {
+		return err
+	}
+
+	// TODO: updatePlusOne isn't good enough to avoid interruption on our infra
+	// nodes.
+	err = updatePlusOne(ctx, cs, p, ssc, vmc, api.AgentPoolProfileRoleInfra)
+	if err != nil {
+		return err
+	}
+
+	err = updatePlusOne(ctx, cs, p, ssc, vmc, api.AgentPoolProfileRoleCompute)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getCount(cs *api.OpenShiftManagedCluster, role api.AgentPoolProfileRole) int {
+	for _, app := range cs.Properties.AgentPoolProfiles {
+		if app.Role == role {
+			return app.Count
+		}
+	}
+
+	panic("invalid role")
+}
+
+func listVMs(ctx context.Context, cs *api.OpenShiftManagedCluster, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) ([]compute.VirtualMachineScaleSetVM, error) {
+	vmPages, err := vmc.List(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), "", "", "")
+	if err != nil {
+		return nil, err
+	}
+
+	var vms []compute.VirtualMachineScaleSetVM
+	for vmPages.NotDone() {
+		vms = append(vms, vmPages.Values()...)
+
+		err = vmPages.Next()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return vms, nil
+}
+
+// updatePlusOne creates an extra VM, then runs updateInPlace, then removes the
+// extra VM.
+func updatePlusOne(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin, ssc compute.VirtualMachineScaleSetsClient, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) error {
```

Comment:
We need to do some testing. for a cluster in numbers of xx nodes this might take AGES. Maybe for future work `UpdateAZ`? Or batch updates?

Reply:
It is slow.  @pweil- should we put a limit on #vms for PP?

Reply:
I think we need to discuss cluster sizing and number of cluster.  Private preview sounds great unless 5 customers create 100 clusters each.